lib/farscape/plugins.rb

```diff
@@ -2,59 +2,103 @@
 
 module Farscape
 
-  @plugins = []
-  @disabled_plugins = []
-  @middleware_stack = PartiallyOrderedList.new { |m,n| order_middleware(m,n) }
+  @plugins = {}
+  @disabling_rules = []
+  @_middleware_stack = nil
 
   class <<self
 
     attr_reader :plugins
-    attr_reader :disabled_plugins
-    attr_reader :middleware_stack
+    attr_reader :disabling_rules
 
     def register_plugin(options)
-      return false if disabled?(options)
-      @plugins << options
-      add_middleware(options)
-      true
+      @middleware_stack = nil
+      options[:enabled] = enabled?(options)
+      @plugins[options[:name]] = options
+    end
+
+    def register_plugins(a_list)
+      a_list.each { |options| register_plugin(options) }
+    end
+
+    def enabled_plugins
+      @plugins.select { |plugin| @plugins[plugin][:enabled] }
+    end
+
+    def disabled_plugins
+      @plugins.reject { |plugin| @plugins[plugin][:enabled] }
     end
 
     # If the middleware has been disabled by name, return the name
     # Else if by type, return the type.
+    # Else if :default_state was passed in return :default_state
+    def why_disabled(options)
+      maybe = @disabling_rules.map { |hash| hash.select { |k,v| k if v == options[k] } }
+      maybe |= disabled_plugins.map { |k,v| v[:name] }
+      maybe |= [:default_state] if options[:default_state] == :disabled
+      maybe || []
```

Comment:
My comment got lost here, I guess. This line can just read `maybe`. Or you could switch the order of 37 and 38 and remove this line altogether.